ENH Replace react-selectable with react-drag-to-select

emteknetnz opened the pull request:
Issue https://github.com/silverstripe/silverstripe-asset-admin/issues/1528

emteknetnz commented on client/src/containers/Gallery/Gallery.js:
```diff
@@ -548,28 +547,6 @@ class Gallery extends Component {
-   * @param Event e
-   * @returns {boolean}
-   */
-  handleBeginSelection(e) {
```
I'm not sure what these actually did, from what i can tell it didn't actually do anything. It doesn't appear to be necessary going forward.

GuySartorelli replied:
This did exactly what it said it did - if you try to click-and-drag from the "next" or "previous" page button at the bottom of the gallery, without this PR it doesn't start a selection, but with this PR it does.

_Very_ minor difference, but it is a little cleaner if it doesn't try selecting when your click starts on another interactive element.

emteknetnz replied:
Have added back in though changed name to `handleShouldStartSelecting` to reflect what it's used on in `<Selectable>`

GuySartorelli reviewed (changes requested):
This has broken drag-and-drop to upload. This video shows me dragging a folder over the asset admin, but the display is completely wrong.

[Screencast from 18-03-25 16:59:34.webm](https://github.com/user-attachments/assets/c5f24222-1ef5-4a2b-9c13-1e13d0c1ebcc)

Compare with CMS 6 which allows me to drop the file(s) anywhere.

https://github.com/user-attachments/assets/04917149-4019-44f1-bba2-76dadd8d712c

GuySartorelli reviewed (changes requested):
Just noticed as well that if I have only a few files, and I start dragging from _below_ them, the dotted box isn't actually displayed until I drag up and my cursor reaches where the files are.

emteknetnz commented on client/src/components/GalleryItem/draggable.js:
```diff
@@ -10,8 +10,8 @@ export default function draggable(Item) {
       className="gallery-item__draggable"
-      // The `draggable` attribute is needed to tell react-selectable to not start selecting on mousedown
-      draggable
+      // The `data-draggable` attribute is needed to tell react-drag-to-select to not start selecting on mousedown
+      data-draggable="true"
```
In theory it's the library https://github.com/AirLabsTeam/react-drag-to-select/blob/main/src/hooks/useSelectionContainer.tsx#L34

But in practice it didn't seem to work, so I manually implemented this in Selectable::shouldStartSelecting()

I'll update the comment

emteknetnz commented on client/src/containers/Gallery/Gallery.js:
```diff
@@ -65,7 +65,6 @@ class Gallery extends Component {
     this.handleBulkInsert = this.handleBulkInsert.bind(this);
-    this.handleBeginSelection = this.handleBeginSelection.bind(this);
     this.handleGroupSelect = this.handleGroupSelect.bind(this);
     this.handleClearSelection = this.handleClearSelection.bind(this);
```
Yeah, it's passed into `<BulkActionsComponent>` in `Gallery.renderBulkActions()`

emteknetnz commented on client/src/containers/Gallery/Gallery.js:
```diff
@@ -957,14 +934,9 @@ class Gallery extends Component {
-            fixedPosition
+          <Selectable
+            isEnabled={this.props.view === 'tile'}
+            onSelectionChange={this.handleGroupSelect}
```
Didn't realise I had, added in

emteknetnz commented on client/src/containers/Selectable/Selectable.js:
```diff
@@ -0,0 +1,127 @@
+/**
+ * Component that wraps around a list of items to make them selectable by dragging a box around them
+ * and calls a callback when the selection changes
+ * Uses the react-drag-to-select library
```
Updated

GuySartorelli commented on client/src/containers/Gallery/Gallery.js:
```diff
@@ -944,6 +943,8 @@ class Gallery extends Component {
       cssClasses.push('gallery__main--has-opened-item');
     }
 
+    const isEnabled = this.props.view === 'tile' && this.props.type === ACTION_TYPES.ADMIN;
```
```suggestion
    const canSelect = this.props.view === 'tile' && this.props.type === ACTION_TYPES.ADMIN;
```
`isEnabled` out here looks like it's talking about the enabled state of the gallery

emteknetnz replied:
Updated

emteknetnz commented:
> After selecting some items, clicking in an empty space doesn't clear the selection.

Fixed

> This PR creates a horizontal scrolling area if you scroll off to the side - BUT that disappears as soon as you let go of the mouse.

Fixed

> This PR also makes it so that if you click and drag from the pagination number it starts selecting text which it didn't do before.

Kind of weird how this started. I've created a [new PR](https://github.com/silverstripe/silverstripe-admin/pull/1913) to fix this.

> Clicking in a blank area does not clear the text selection. Clicking and dragging to select files after that also selects their text.

Fixed

> It is a little weird that the selection display gets cut off on the left earlier than the left menu, and earlier going upwards than the breadcrumb section.

I think we'll need to leave this as is as I don't know how to resolve this. IMO it's a pretty minor aesthetic issue that doesn't affect actual functionality at all.

GuySartorelli commented:
>> It is a little weird that the selection display gets cut off on the left earlier than the left menu, and earlier going upwards than the breadcrumb section.
> I think we'll need to leave this as is as I don't know how to resolve this. IMO it's a pretty minor aesthetic issue that doesn't affect actual functionality at all.

It is really disorienting for me - but if there's not much we can do then there's not much we can do.

My one thought is - what if we made the `.gallery__outer` element the parent of the scrollable? 
![image](https://github.com/user-attachments/assets/a765201c-eb88-49cb-bb77-a62205d55560)
If that's doable without much effort, I think it would resolve it, because then the boundaries for where it stops rendering are marked with clear lines rather than invisible ones.

If that's not doable or is considerable effort, I'm okay with not doing it.

emteknetnz commented:
Yeah that seemed to work, the top and left boundaries aren't dotted lines unfortunately, though there's less "cutting off" when you drag above the area

GuySartorelli commented:
That's way better🎉

GuySartorelli reviewed (approved):
LGTM, works as expected
